### PR TITLE
Improve iframe switching via CDP

### PR DESCRIPTION
**Pull request by LavrovArtem**

### CODE REVIEW

File: src/browser/provider/built-in/dedicated/chrome/cdp-client/client-function-executor.ts [hunk 1]
```diff
@@ -127,8 +149,11 @@ export default class ClientFunctionExecutor {
         return JSON.parse(result!.value); // eslint-disable-line @typescript-eslint/no-non-null-assertion
     }
 
-    public async executeSelector (Runtime: RuntimeApi, command: ExecuteSelectorCommand, callsite: CallsiteRecord, selectorTimeout: number): Promise<object> {
-        const expression = `${PROXYLESS_SCRIPT}.executeSelectorCommand(${JSON.stringify(command)}, ${selectorTimeout}, ${Date.now()});`;
+    public async executeSelector <T extends SelectorSnapshotArgs | SelectorNodeArgs> (args: T): Promise<T extends SelectorNodeArgs ? string : object> {
+        const { Runtime, command, callsite, selectorTimeout, errTypes } = args;
+
```
Comment (miherlosev) [inline src/browser/provider/built-in/dedicated/chrome/cdp-client/client-function-executor.ts:154]: Move this expression to constant and use it below.
```js
const returnNode = 'DOM' in args;
```

File: src/browser/provider/built-in/dedicated/chrome/cdp-client/client-function-executor.ts [hunk 2]
```diff
@@ -138,18 +163,26 @@ export default class ClientFunctionExecutor {
         if (exceptionDetails)
             ClientFunctionExecutor._throwException(exceptionDetails, command, callsite);
 
-        return JSON.parse(result!.value); // eslint-disable-line @typescript-eslint/no-non-null-assertion
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        return 'DOM' in args ? result!.objectId : JSON.parse(result!.value);
+    }
+
+    public async executeSelectorAndGetNode (args: SelectorNodeArgs): Promise<DOM.Node> {
```
Comment (miherlosev) [inline src/browser/provider/built-in/dedicated/chrome/cdp-client/client-function-executor.ts:170]: `public async getNode (args: SelectorNodeArgs): Promise<DOM.Node> {`

File: src/browser/provider/built-in/dedicated/chrome/cdp-client/client-function-executor.ts [hunk 3]
```diff
@@ -6,12 +6,11 @@ import EvaluateRequest = Protocol.Runtime.EvaluateRequest;
...
-    CannotObtainInfoForElementSpecifiedBySelectorError,
-    ClientFunctionExecutionInterruptionError, DomNodeClientFunctionResultError,
-    InvalidSelectorResultError, UncaughtErrorInClientFunctionCode,
-    UncaughtErrorInCustomDOMPropertyCode,
-} from '../../../../../../shared/errors';
+import DOMApi = ProtocolProxyApi.DOMApi;
+import ExecutionContextCreatedEvent = Protocol.Runtime.ExecutionContextCreatedEvent;
+import ExecutionContextDestroyedEvent = Protocol.Runtime.ExecutionContextDestroyedEvent;
+import DOM = Protocol.DOM;
+import * as Errors from '../../../../../../shared/errors';
```
Comment (miherlosev) [inline src/browser/provider/built-in/dedicated/chrome/cdp-client/client-function-executor.ts:13]: Use `SharedErrors` alias here and below.

File: src/browser/provider/built-in/dedicated/chrome/cdp-client/index.ts [hunk 4]
```diff
@@ -332,26 +334,50 @@ export class BrowserClient {
...
+                notFound:  Errors.CannotObtainInfoForElementSpecifiedBySelectorError.name,
+                invisible: Errors.CannotObtainInfoForElementSpecifiedBySelectorError.name,
+            },
+
+            command, callsite, selectorTimeout,
+        });
     }
 
-    public async switchToIframe (): Promise<void> {
+    public async switchToIframe (command: SwitchToIframeCommand, callsite: CallsiteRecord, selectorTimeout: number): Promise<void> {
```
Comment (miherlosev) [inline src/browser/provider/built-in/dedicated/chrome/cdp-client/index.ts:348]: Use a single object parameter here.
`public async switchToIframe ({ command: SwitchToIframeCommand, callsite: CallsiteRecord, selectorTimeout: number }): Promise<void> {`

File: src/browser/provider/built-in/dedicated/chrome/cdp-client/index.ts [hunk 5]
```diff
@@ -332,26 +334,50 @@ export class BrowserClient {
...
+        });
 
         if (!node.frameId)
-            return;
+            throw new Errors.ActionElementNotIframeError(callsite);
 
         this._clientFunctionExecutor.setCurrentFrameId(node.frameId);
+
+        // TODO:
+        // return this._ensureChildIframeDriverLink(nativeMethods.contentWindowGetter.call(iframe),
```
Comment (miherlosev) [inline src/browser/provider/built-in/dedicated/chrome/cdp-client/index.ts:379]: Remove this development artifacts.

File: src/client/proxyless/index.ts [hunk 6]
```diff
@@ -22,15 +22,18 @@ Object.defineProperty(window, '%proxyless%', {
...
-        executeSelectorCommand: function (command: ExecuteSelectorCommand, selectorTimeout: number, startTime: number) {
-            const createError = command.needError
-                ? (fn: FnInfo | null) => new CannotObtainInfoForElementSpecifiedBySelectorError(null, fn)
-                : null;
+        executeSelectorCommand: function (command: ExecuteSelectorCommand, selectorTimeout: number, startTime: number,
+            returnNode: boolean, errTypes: { notFound: string; invisible: string }) {
 
-            const selectorExecutor = new SelectorExecutor(command, selectorTimeout, startTime, createError, createError);
+            // @ts-ignore
+            const createNotFoundError    = command.needError ? (fn: FnInfo | null) => new Errors[errTypes.notFound](null, fn) : null;
```
Comment (miherlosev) [inline src/client/proxyless/index.ts:29]: Refactor in such a way:
```js
const createNotFoundError    = command.needError ? createErrorFn(errTypes.notFound): null;
const createIsInvisibleError = command.needError ? createErrorFn(errTypes.invisible): null;
```

**Review by miherlosev — CHANGES_REQUESTED** [review]

File: src/browser/provider/built-in/dedicated/chrome/cdp-client/index.ts (hunk 4 above)
Comment (LavrovArtem) [reply src/browser/provider/built-in/dedicated/chrome/cdp-client/index.ts:348]: I think it's unnecessary, because it's only 3 parameters.

**Review by miherlosev — APPROVED** [review]

**Review by AndreyBelym — APPROVED** [review]

**Comment by AndreyBelym** [comment]
Please do not merge until v1.16.0 is out.

**Pull request merged**